History performance improvements for single-entity requests

OverloadUT opened the pull request:
## Description:
Drastically improves performance when fetching the history of a single entity. It accomplishes this by doing an entirely different query for fetching the "synthetic first datapoint" when only a single entity_id is being searched for.

There is still a ton of room for improvement when it comes to multi-entity history fetching which I am continuing to investigate, but this first change makes a massive improvement for all single entity popups.

The performance change in some of my tests shows this change to be about 300x faster (30 seconds to 0.1 seconds!) (There is no theoretical upper limit due to the issue with multi-entity queries getting worse the longer the current recorder run has been going.)

**Related issue (if applicable):** fixes n/a

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** none

## Example entry for `configuration.yaml` (if applicable): none

## Checklist:

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works. *(not needed; existing tests cover the code already)*

OverloadUT committed:
Bugfix: remove superfluous domain filter

This filter is already applied later in the function by the `filters` object, where it is conditionally applied when appropriate. This fixes the problem where we get a domain filter even when searching for a single entity_id, which needlessly harms the query's performance.

OverloadUT committed:
Performance: build different query when only getting single entity

When querying the history of a single entity, we can use an entirely different method for the "synthetic zero data point" by simply sorting by date and doing a LIMIT 1. This performs thousands of times better than the multi-entity query when the current recorder_run has been going for a while.

OverloadUT committed:
Add entity_id filter to single-entity request

The entity_id filter was handled inside the `filters.apply` logic which is used in most cases, BUT didn't work when no `filters` was passed in to the method. Now it'll work even if no `filters` object is passed in.

OverloadUT committed:
Fix linting errors in history.py

fanaticDavid commented:
Sooo looking forward to this going live, if your results are anything to go by!

Oh, and by the way: going from 30 seconds to 0.1 seconds is 300 times faster!! 😁

OverloadUT commented:
@fanaticDavid: Updated the description! 😄

azogue commented:
Testing it now, wow, thanks! I'll comment on my impressions.

cgarwood commented:
Super excited for this.

FWIW it looks like the comment around line 140 was truncated:
`+            # We have more than one entity to look at, so we need to do a much`

armills commented on homeassistant/components/history.py:
```diff
@@ -119,19 +119,39 @@ def get_states(hass, utc_point_in_time, entity_ids=None, run=None,
+            # have a single entity id
+            most_recent_state_ids = session.query(
+                States.state_id.label('max_state_id')
+            ).filter(
```
Can we keep `(States.created >= run.start)` here? It would be nice to keep it as a safeguard against weird conditions. I don't believe it should affect the query performance. You could also try including it as part of a `between` condition, but I'm pretty sure it will be optimized to the same thing either way.

If it has a large impact I'm OK dropping it.

OverloadUT replied:
I just tested it in MySQL and I am seeing no performance impact to include it.

My preference would be to not do this unless we can think of an edge case where it will protect against something. I get the desire, but I worry we'd be sacrificing future readability of the code for a guard that may not be necessary...

armills replied:
I suppose I'm OK leaving it out. I just get worried about weird regressions. 😃 I guess it was mainly for the case of not knowing which `entity_ids` you wanted.

armills commented on homeassistant/components/history.py:
```diff
@@ -119,19 +119,39 @@ def get_states(hass, utc_point_in_time, entity_ids=None, run=None,
+                func.max(States.state_id).label('max_state_id')
+            ).filter(
+                (States.created >= run.start) &
+                (States.created < utc_point_in_time))
```
We lost `(~States.domain.in_(IGNORE_DOMAINS)))` here.

OverloadUT replied:
That was actually intentional (I put it in its own commit to call it out) as it was causing the domain filter to appear twice in every query, due to the fact that the `filters.apply` process applies the IGNORE_DOMAINS itself. This was the only method that had this double-logic so I figured it was safe to remove it and cleans things up a tad.

Interestingly (and this is a topic for a refactor), it appears that the `filters.apply` logic *always* happens when these methods are used in production. The unit tests are the only code that skips the scaffolding and goes straight to the methods without supplying a `filters` object.

armills replied:
Ah, I didn't look at the individual commits. Good catch!

It looks to me like we're [using it here](https://github.com/home-assistant/home-assistant/blob/f43db3c615db00c578b03014136e5ddd2289e830/homeassistant/helpers/restore_state.py#L43) without the filters. IMO we should drop it from `filters.apply`, since it seems out of place there, and we want to continue using it for `restore_state`.

OverloadUT replied:
Ah ha! PyCharm's "find usages" feature failed me! I should have done a full text search. Okay, I'll change it around to make the logic live in the method and remove it from `filters.apply`

OverloadUT replied:
Alright, I put the domain filter back. I decided to not to drop it from filters.apply in this change, because I'm not sure which other methods would need to gain that logic to ensure everything stays the same.

armills replied:
OK, that's fine. It's definitely good to know it's getting double applied though, we'll want to fix that in the future.

emlove reviewed (commented):
Excellent work!

This is probably indicative of refactoring we should be doing in the future, but the branch is well commented.

OverloadUT committed:
Undo removal of domain filter

Putting back the domain filter that was removed in 76a6371705dcd57483e55dcc03435ae867c184d2 - there are use-cases where get_states is called without a filter object, so we need the domain filter to work in those cases as well.

diplix commented:
🥂

OverloadUT committed:
Fix truncated comment

armills commented:
This is going to be a huge usability improvement. Great work! Looking forward to more contributions! 🎉
